cpp/src/arrow/flight/sql/odbc/odbc_api.cc

```diff
@@ -223,4 +243,120 @@ SQLRETURN SQLSetEnvAttr(SQLHENV env, SQLINTEGER attr, SQLPOINTER valuePtr,
     }
   });
 }
+
+SQLRETURN SQLDriverConnectW(SQLHDBC conn, SQLHWND windowHandle,
+                            SQLWCHAR* inConnectionString,
+                            SQLSMALLINT inConnectionStringLen,
+                            SQLWCHAR* outConnectionString,
+                            SQLSMALLINT outConnectionStringBufferLen,
+                            SQLSMALLINT* outConnectionStringLen,
+                            SQLUSMALLINT driverCompletion) {
+  // TODO: Implement FILEDSN and SAVEFILE keywords according to the spec
+  // https://github.com/apache/arrow/issues/46449
+
+  std::cout << "-AL- SQLDriverConnectW called\n";
+
+  using driver::odbcabstraction::Connection;
+  using ODBC::ODBCConnection;
+
+  return ODBCConnection::ExecuteWithDiagnostics(conn, SQL_ERROR, [=]() {
+    ODBCConnection* connection = reinterpret_cast<ODBCConnection*>(conn);
+    std::string connection_string =
+        ODBC::SqlWcharToString(inConnectionString, inConnectionStringLen);
+    Connection::ConnPropertyMap properties;
+    std::string dsn =
+        ODBCConnection::getPropertiesFromConnString(connection_string, properties);
+
+    std::vector<std::string> missing_properties;
+
+    // TODO: Implement SQL_DRIVER_COMPLETE_REQUIRED in SQLDriverConnectW according to the
+    // spec https://github.com/apache/arrow/issues/46448
+#if defined _WIN32 || defined _WIN64
+    if (driverCompletion == SQL_DRIVER_PROMPT ||
+        ((driverCompletion == SQL_DRIVER_COMPLETE ||
+          driverCompletion == SQL_DRIVER_COMPLETE_REQUIRED) &&
+         !missing_properties.empty())) {
+      // TODO: implement driverCompletion behavior to display connection window.
+    }
+#endif
+
+    connection->connect(dsn, properties, missing_properties);
+
+    // Copy connection string to outConnectionString after connection attempt
+    return ODBC::GetStringAttribute(true, connection_string, true, outConnectionString,
+                                    outConnectionStringBufferLen, outConnectionStringLen,
+                                    connection->GetDiagnostics());
+  });
+}
+
+SQLRETURN SQLConnectW(SQLHDBC conn, SQLWCHAR* dsnName, SQLSMALLINT dsnNameLen,
+                      SQLWCHAR* userName, SQLSMALLINT userNameLen, SQLWCHAR* password,
+                      SQLSMALLINT passwordLen) {
+  std::cout << "-AL- SQLConnectW called\n";
+  using driver::flight_sql::FlightSqlConnection;
+  using driver::flight_sql::config::Configuration;
+  using ODBC::ODBCConnection;
+
+  using ODBC::SqlWcharToString;
+
+  ODBCConnection* connection = reinterpret_cast<ODBCConnection*>(conn);
+
+  return ODBCConnection::ExecuteWithDiagnostics(conn, SQL_ERROR, [=]() {
+    std::string dsn = SqlWcharToString(dsnName, dsnNameLen);
+
+    Configuration config;
+    config.LoadDsn(dsn);
+
+    if (userName) {
+      std::string uid = SqlWcharToString(userName, userNameLen);
+      config.Set(FlightSqlConnection::UID, uid);
```

Comment:
Micro-optimization. Add a move-version of config.Set() (eg Set(const std::string& key, std::string&& value)) and call it Emplace(). Then use config.Emplace(key, std::move(value)) to avoid copying.